one-time pass thru error-prone NarrowingCompoundAssignment check

### Description

A great overflow issue: https://errorprone.info/bugpattern/NarrowingCompoundAssignment

javac is getting its own checker for this in version 20 so eventually we can switch to that one. But for now, i'd like to try to enable the error-prone version.

in a nutshell, compound assignment operators are totally lenient:
```
jshell> byte x = 0;
x ==> 0

jshell> x = x + Long.MAX_VALUE;
|  Error:
|  incompatible types: possible lossy conversion from long to byte
|  x = x + Long.MAX_VALUE;
|      ^----------------^

jshell> x += Long.MAX_VALUE;
$2 ==> -1
```

Here's a log of the output from the new check: [overflow.log](https://github.com/apache/lucene/files/10014404/overflow.log)

I think it is too noisy to worry about fixing all the code. I plan to do a single pass through it and fix anything that looks "real". For example places where, `int` accumulator is accidentally used to sum up `long` values with `+=`. In these cases the wrong type was used but java threw away all safety thanks to `+=`.

Example: 
```
/home/rmuir/workspace/lucene/lucene/core/src/java/org/apache/lucene/codecs/KnnVectorsWriter.java:141: warning: [NarrowingCompoundAssignment] Compound assignments from long to int hide lossy casts
        totalCost += sub.values.cost();
                  ^
    (see https://errorprone.info/bugpattern/NarrowingCompoundAssignment)
  Did you mean 'totalCost = (int) (totalCost + sub.values.cost());'?
```
In this case KnnVectorsWriter `totalCost` should simply be a `long` instead of `int`. 

Another similar one:
```
/home/rmuir/workspace/lucene/lucene/core/src/java/org/apache/lucene/util/OfflineSorter.java:302: warning: [NarrowingCompoundAssignment] Compound assignments from long to int hide lossy casts
        sortInfo.lineCount += part.count;
```
In this case we should change `sortInfo.lineCount` to be a `long` instead of `int`.

--- a/lucene/core/src/java/org/apache/lucene/codecs/KnnVectorsWriter.java
+++ b/lucene/core/src/java/org/apache/lucene/codecs/KnnVectorsWriter.java
@@ -108,7 +108,7 @@ public abstract class KnnVectorsWriter implements Accountable, Closeable {
   protected static class MergedVectorValues extends VectorValues {
     private final List<VectorValuesSub> subs;
     private final DocIDMerger<VectorValuesSub> docIdMerger;
-    private final int cost;
+    private final long cost;
     private final int size;
 
     private int docId;
@@ -136,7 +136,8 @@ public abstract class KnnVectorsWriter implements Accountable, Closeable {
         throws IOException {
       this.subs = subs;
       docIdMerger = DocIDMerger.of(subs, mergeState.needsIndexSort);
-      int totalCost = 0, totalSize = 0;
+      long totalCost = 0;
+      int totalSize = 0;
       for (VectorValuesSub sub : subs) {
         totalCost += sub.values.cost();
         totalSize += sub.values.size();

--- a/lucene/core/src/java/org/apache/lucene/codecs/KnnVectorsWriter.java
+++ b/lucene/core/src/java/org/apache/lucene/codecs/KnnVectorsWriter.java
@@ -108,7 +108,6 @@ public abstract class KnnVectorsWriter implements Accountable, Closeable {
   protected static class MergedVectorValues extends VectorValues {
     private final List<VectorValuesSub> subs;
     private final DocIDMerger<VectorValuesSub> docIdMerger;
-    private final long cost;
     private final int size;
 
     private int docId;
@@ -136,13 +135,10 @@ public abstract class KnnVectorsWriter implements Accountable, Closeable {
         throws IOException {
       this.subs = subs;
       docIdMerger = DocIDMerger.of(subs, mergeState.needsIndexSort);
-      long totalCost = 0;
       int totalSize = 0;
       for (VectorValuesSub sub : subs) {
-        totalCost += sub.values.cost();
         totalSize += sub.values.size();
       }
-      cost = totalCost;
       size = totalSize;
       docId = -1;
     }
@@ -185,7 +181,7 @@ public abstract class KnnVectorsWriter implements Accountable, Closeable {
 
     @Override
     public long cost() {
-      return cost;
+      return size;
     }
 
     @Override

--- a/lucene/core/src/java/org/apache/lucene/search/PointRangeQuery.java
+++ b/lucene/core/src/java/org/apache/lucene/search/PointRangeQuery.java
@@ -415,7 +415,7 @@ public abstract class PointRangeQuery extends Query {
           BiFunction<byte[], byte[], Relation> nodeComparator,
           Predicate<byte[]> leafComparator)
           throws IOException {
-        final int[] matchingNodeCount = {0};
+        final long[] matchingNodeCount = {0};
         // create a custom IntersectVisitor that records the number of leafNodes that matched
         final IntersectVisitor visitor =
             new IntersectVisitor() {
@@ -446,7 +446,7 @@ public abstract class PointRangeQuery extends Query {
       }
 
       private void pointCount(
-          IntersectVisitor visitor, PointValues.PointTree pointTree, int[] matchingNodeCount)
+          IntersectVisitor visitor, PointValues.PointTree pointTree, long[] matchingNodeCount)
           throws IOException {
         Relation r = visitor.compare(pointTree.getMinPackedValue(), pointTree.getMaxPackedValue());
         switch (r) {

--- a/lucene/core/src/java/org/apache/lucene/util/OfflineSorter.java
+++ b/lucene/core/src/java/org/apache/lucene/util/OfflineSorter.java
@@ -141,7 +141,7 @@ public class OfflineSorter {
     /** number of partition merges */
     public int mergeRounds;
     /** number of lines of data read */
-    public int lineCount;
+    public long lineCount;
     /** time spent merging sorted partitions (in milliseconds) */
     public final AtomicLong mergeTimeMS = new AtomicLong();
     /** time spent sorting data (in milliseconds) */

